File: problems/06_linked_list/0287_find_the_duplicate_number/solution.py
"""
Problem: 0287. Find the Duplicate Number
Difficulty: Medium
URL: https://leetcode.com/problems/find-the-duplicate-number/

解法演進：
1) HashSet O(n)/O(n)。
2) Floyd 找環起點 O(n)/O(1)（最佳）。

最佳解核心思路：
- i -> nums[i] 構成函式圖，n+1 個值映到 [1,n] 必有環，入口即重複值。
"""

class Solution:
    def findDuplicate(self, nums: list[int]) -> int:
        pass

    def findDuplicate_brute(self, nums: list[int]) -> int:
        pass


if __name__ == "__main__":
    solution = Solution()

    # Test 1
    # result = solution.solve(...)
    # assert result == expected

    print("All tests passed!")

File: problems/06_linked_list/0287_find_the_duplicate_number/solution.dart
// Problem: 0287. Find the Duplicate Number
// Difficulty: Medium

int findDuplicate(List<int> nums) {
  var slow = 0, fast = 0;
  while (true) {
    slow = nums[slow];
    fast = nums[nums[fast]];
    if (slow == fast) break;
  }
  slow = 0;
  while (slow != fast) { slow = nums[slow]; fast = nums[fast]; }
  return slow;
}

void _expect(int a, int e, String msg) {
  if (a != e) throw StateError('FAIL $msg: $a vs $e');
}

void main() {
  _expect(findDuplicate([1, 3, 4, 2, 2]), 2, 'ex1');
  _expect(findDuplicate([3, 1, 3, 4, 2]), 3, 'ex2');
  _expect(findDuplicate([1, 1]), 1, 'two');
  _expect(findDuplicate([2, 2, 2, 2, 2]), 2, 'all same');
  _expect(findDuplicate([1, 4, 4, 2, 4]), 4, 'multi');
  print('All tests passed!');
}
